### PR TITLE
Rating backend

### LINKED ISSUE #29: Create a backend POST to rate a player
**Description**:  
Create a backend POST to rate a player

### LINKED ISSUE #31: Create backend POST to comment on user profiles
**Description**:  
Create backend POST to comment on user profiles

### LINKED ISSUE #93: Edit rating backend
**Description**:  
user can't rerate a profile but can edit rating.

### PR DESCRIPTION
**Pull request by humayraR**
AddRating: 
* can rate/comment user profile if logged in
* will update if the rating by rater already exists on that profile.

getUserRatings:
* get list of all ratings for a user profile

getUserAvgRatings:
* get the average of all ratings and the list of all comments(with comment time, commenterID)

getUserRatingbyRater:
* get rating for user by particular rater

Tests Complete.

Closes #29 
Closes #31 
Closes #93

### CODE REVIEW

File: gamersnet_backend/persistence/ratings.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * @param {*} fun is the user fun to play with?
+ * @param {*} playAgain would the user wanna play again?
+ * @param {*} comment any comment on the user
+ * @returns 
+ */
+async function addRatingsDB(userID, raterID, strength, punctuality, friendliness, fun, playAgain, comment) {
+  // wait for db connection and get ratings collection
+  let db = await MongoDB.open();
+
+  let posts = db.collection('ratings');
```
Comment (jayKhakhariya) [inline gamersnet_backend/persistence/ratings.js]: change the variable name 'posts' to ratings

File: gamersnet_backend/routes/ratings/addRatings.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        cookie = cookie.split('=')[1];
+        loggedIn = await verifyUserLoggedIn(cookie);
+    }
+
+    if (loggedIn && verifyBody) {
+        let tokenDocument = await getUserIDFromToken(cookie);
+        let raterID = tokenDocument.userID;
+
+        // users can't rate their own profiles.
+        if(!raterID.equals(ObjectID(userID))) {
```
Comment (jayKhakhariya) [inline gamersnet_backend/routes/ratings/addRatings.js]: declare a variable id = ObjectID(userID) as it is used multiple times

**Review by jayKhakhariya — CHANGES_REQUESTED** [review]
Looks good to me apart from minor changes I suggested

**Comment by humayraR** [comment]
> Looks good to me apart from minor changes I suggested

fixed

File: gamersnet_backend/__tests__/ratings/testAddRatings.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        return request(app).post('/ratings/addRatings')
+        .set('Cookie', '')//clear cookie and reset
+        .set('Cookie', 'token=user2_token')
+        .query({userID: user1ID.toHexString()})
+        .send({
+            strength: 5, 
+            punctuality: 5, 
+            friendliness: 4,
+            fun: 4,
+            playAgain: true,
```
Comment (lucashuy) [inline gamersnet_backend/__tests__/ratings/testAddRatings.js:71]: Here you are setting `playAgain` to true

File: gamersnet_backend/__tests__/ratings/testGetUserRatings.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+//the scheduled time of this game is still due
+let rating1 = {
+    userID: ObjectID(),// intended to link to existing users in db
+    raterID: ObjectID(),
+    strength: 5,
+    punctuality: 4, 
+    friendliness: 4, 
+    fun: 4,
+    playAgain: "yes",
```
Comment (lucashuy) [inline gamersnet_backend/__tests__/ratings/testGetUserRatings.js]: but there here you are setting it to `"yes"` in the next test

File: gamersnet_backend/routes/ratings/getUserRatings.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        let playAgain = 0;
+        let comments = [];
+        
+        let i = 0;
+        for(i = 0; i < results.length; i++){
+            strength += results[i].strength;
+            punctuality += results[i].punctuality; 
+            friendliness += results[i].friendliness; 
+            fun += results[i].fun;
+            if(results[i].playAgain === 'yes')
```
Comment (lucashuy) [inline gamersnet_backend/routes/ratings/getUserRatings.js]: and here in the route callback you are checking for the string `"yes"`.

Consider changing this to a boolean as it was in the first test

File: gamersnet_backend/__tests__/ratings/testAddRatings.js (hunk 3 above)
Comment (humayraR) [reply gamersnet_backend/__tests__/ratings/testAddRatings.js:71]: Changed all to boolean type

**Comment by lucashuy** [comment]
LGTM

**Review by lucashuy — DISMISSED** [review]

**Review by lucashuy — APPROVED** [review]
LGTM

**Review by jayKhakhariya — APPROVED** [review]
Looks good!

**Pull request merged**